application/ui/components/input/custom-object.jsx

```diff
@@ -0,0 +1,206 @@
+/** @jsx React.DOM */
+'use strict';
+
+var _             = require('underscore');
+var React         = require('react');
+var ObjectParameter;
+
+module.exports = ObjectParameter = React.createClass({
+
+    displayName : 'ObjectParameter',
+
+    getInitialState : function()
+    {
+        return {
+            values : []
+        };
+    },
+
+    getValue : function()
+    {
+        return this.state.values;
+    },
+
+    getDefaultValue : function()
+    {
+        return {
+            key: '',
+            value: '',
+            inputType: 'string'
+        };
+    },
+
+    getInputs : function()
+    {
+        var inputs   = [],
+            instance = this;
+
+        _.each(this.state.values, function (value, index) {
+            inputs.push(instance.getInput(value, index));
+        });
+
+        return inputs;
+    },
+
+    getInput : function(prop, index)
+    {
+        var field, selectType, instance = this;
+
+        if (prop.inputType !== 'object') {
+            field = (
+                <input
+                    className    = 'array-input'
+                    value        = {prop.value}
+                    onChange     = {_.partial(instance.updateField, index, 'value')}
+                />
+            );
+        } else {
+            field = (
+                <table>
+                    <ObjectParameter value={prop.value} index={index} onChange={_.partial(instance.updateField, index, 'value')}/>
+                </table>
+            );
+        }
+
+        var rmCallback = function () {
+            instance.removeField(index);
+        };
+
+        if (prop.inputType === 'object') {
+            selectType = 'object';
+        } else {
+            selectType = (
+                <select
+                    className='select'
+                    onChange={_.partial(instance.updateField, index, 'inputType')}
+                    value={prop.inputType}
+                >
+                    <option value={'string'}>String</option>
+                    <option value={'number'}>Number</option>
+                </select>
+            );
+        }
+
+        return (
+            <tr key={index}>
+                <td>
+                    <a className='button field-button--remove' onClick={rmCallback}>–</a>
+                </td>
+                <td>
+                    <label>Key: <input
+                        className    = 'array-input'
+                        value        = {prop.key}
+                        onChange     = {_.partial(instance.updateField, index, 'key')}
+                    />
```

Comment:
There's also a `TextInput` component in `ui/components/input` that abstracts away the whole "getting the value out of the event object" piece, so the `updateField` method could be made shorter.